examples/dataflow-python-examples/streaming-examples/nlp-sentiment-analysis-pipeline/src/beam/pipeline.py

```diff
@@ -0,0 +1,413 @@
+
+    def process(self, tweet_tuple, tweet_index=beam.DoFn.StateParam(TWEET_INDEX), doc_index=beam.DoFn.StateParam(DOC_INDEX), window=beam.DoFn.WindowParam):
+        time_value, tweet_text = tweet_tuple
+
+        tweet_index.add([time_value, 1])
+
+        if tweet_index.read()[time_value] > 500:
+            tweet_index.clear()
+            doc_index.add([time_value, 1])
+
+        window_start = int(window.start.to_utc_datetime().timestamp())
+        doc_index = doc_index.read().get(time_value, 0)
+
+        modified_key = "{time_value}+{window_start}+{doc_index}".format(time_value=time_value, window_start=window_start, doc_index=doc_index)
+        yield (modified_key, tweet_text)
+
+
+class NlpClient_Batched(beam.DoFn):
+    """Gets sentiment score for each tweet via batching.
+
+    This class batches each tweet into a document. There is
+    one document per window per key. In other words, if there
+    exists 3 dates within a window, there are 3 documents made.
+    The document approach is meant to drastically reduce the
+    number of API calls to the Natural Language API.
+
+    Attributes:
+      language: Language of tweets.
+      client: NLP API client instance.
+    """
+    def __init__(self, client=None):
+        self.language = "en"
+        self.client = client
+
+    def setup(self): # NOTE: setup() is only supported with Dataflow Runner or Direct Runner (Apache Beam 2.22.0)
+        if self.client is None:
+            self.client = language.LanguageServiceClient() 
+
+    def process(self, modified_tweet_tuple):
+        if self.client is None:
+            self.client = language.LanguageServiceClient()
+
+        (modified_key, text_iterable) = modified_tweet_tuple
+
+        time_value = modified_key[:modified_key.index("+")]
+
+        file_name = './' + modified_key + ".txt"
+        writer = filesystems.FileSystems.create(file_name)
+        punctuation = ['.', '?', '!']
+        for tweet in text_iterable:
+            tweet = tweet.replace('...', ' ').replace('\n', ' ')
+            for p in punctuation:
+                tweet = tweet.replace(p, ' ')
+            writer.write((tweet + ".\n").encode('utf-8'))
+
+        writer.close()
+
+        annotations = self.analyze_document(file_name)
+        for tweet in annotations.sentences:
+            yield (time_value, tweet.sentiment.score)
+
+        filesystems.FileSystems.delete([file_name])
+
+    @retry.Retry(predicate=retry.if_exception_type(exceptions.ResourceExhausted), initial=10.0, maximum=100.0, multiplier=2.0)
+    def analyze_document(self, file_name):
+        with open(file_name, 'r+') as review_file:
+            content = review_file.read()
+
+        document = types.Document(
+            content=content,
+            type=enums.Document.Type.PLAIN_TEXT,
+            language=self.language
+        )
+
+        annotations = self.client.analyze_sentiment(document)
+        return annotations
+
+
+class NlpClient_ElementWise(beam.DoFn):
```

Comment:
Seem like you should have am abstract base class for NlpClient the DRY things up and have concrete base class for batched and elementwise.